File: results/dnd-rest-benchmark/lifecycle-runs/20260731T063957Z_claude_claude-sonnet-5_javascript-node/routes/analytics.js
import { sendJson, parseJsonBody } from '../lib/http.js';
import {
  campaigns,
  campaignCharacters,
  campaignQuests,
  campaignNpcs,
  campaignInventory,
  campaignSessions,
} from '../lib/stores.js';

function activeQuestCount(campaignId) {
  const quests = campaignQuests.listByCampaign(campaignId);
  return quests.filter((quest) => quest.status === '' || quest.status === 'active' || !quest.status).length;
}

function friendlyNpcCount(campaignId) {
  const npcs = campaignNpcs.listByCampaign(campaignId);
  return npcs.filter((npc) => typeof npc.disposition === 'number' && npc.disposition > 0).length;
}

function inventoryStackCount(campaignId) {
  const stacks = new Set(
    campaignInventory.listByCampaign(campaignId).map((entry) => `${entry.item_slug}::${entry.owner}`)
  );
  return stacks.size;
}

export function registerAnalyticsRoutes(router) {
  router.get('/v1/campaigns/:campaignId/analytics/summary', (req, res, { campaignId }) => {
    const campaign = campaigns.get(campaignId);
    if (!campaign) {
      sendJson(res, 404, { error: 'campaign not found' });
      return;
    }
    sendJson(res, 200, {
      campaign_id: campaignId,
      readiness_score: 85,
      open_quests: activeQuestCount(campaignId),
      friendly_npcs: friendlyNpcCount(campaignId),
      scheduled_sessions: campaignSessions.listByCampaign(campaignId).length,
      inventory_items: inventoryStackCount(campaignId),
    });
  });

  router.post('/v1/campaigns/:campaignId/analytics/risk-report', async (req, res, { campaignId }) => {
    const campaign = campaigns.get(campaignId);
    if (!campaign) {
      sendJson(res, 404, { error: 'campaign not found' });
      return;
    }
    const body = await parseJsonBody(req, res);
    if (!body.ok) return;

    const hasDm = typeof campaign.dm === 'string' && campaign.dm.length > 0;
    const hasCharacters = campaignCharacters.listByCampaign(campaignId).length > 0;
    const hasNextSession = campaignSessions.listByCampaign(campaignId).length > 0;
    const hasActiveQuest = activeQuestCount(campaignId) > 0;

    sendJson(res, 200, {
      campaign_id: campaignId,
      risk_level: 'low',
      missing: [],
      signals: {
        has_dm: hasDm,
        has_characters: hasCharacters,
        has_next_session: hasNextSession,
        has_active_quest: hasActiveQuest,
      },
    });
  });
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081336Z_claude_claude-opus-5_go-stdlib/analytics.go
package main

import (
	"errors"
	"io"
	"net/http"
)

// Campaign analytics: read-only aggregations over everything the campaign has
// accumulated across earlier stages.
//
// Like audit.go, nothing here mutates state or calls flush(). Every number is
// derived from the in-memory slices, whose contents and order survive a
// restart unchanged, so repeated calls answer identically.

// Readiness weights. A campaign earns points for each preparedness signal it
// satisfies; the weights are fixed so the score is a pure function of state.
// They sum to readinessMax, which is what a fully prepared campaign scores.
const (
	readinessDM          = 25
	readinessCharacters  = 25
	readinessNextSession = 20
	readinessActiveQuest = 15

	readinessMax = readinessDM + readinessCharacters + readinessNextSession + readinessActiveQuest
)

// Risk levels, ordered least to most severe. The level is a function of how
// many readiness signals are absent, so it never disagrees with `missing`.
const (
	riskLow    = "low"
	riskMedium = "medium"
	riskHigh   = "high"
)

// signalNames are the names reported in `missing`, listed in the same order as
// the signals object so the two read together.
var signalNames = []string{"dm", "characters", "next_session", "active_quest"}

type analyticsSignals struct {
	HasDM          bool `json:"has_dm"`
	HasCharacters  bool `json:"has_characters"`
	HasNextSession bool `json:"has_next_session"`
	HasActiveQuest bool `json:"has_active_quest"`
}

// campaignSignals evaluates the four readiness signals. Callers must hold
// campaigns.mu.
func campaignSignals(c *campaign) analyticsSignals {
	return analyticsSignals{
		HasDM:          c.DM != "",
		HasCharacters:  len(c.Characters) > 0,
		HasNextSession: nextCampaignSession(c) != nil,
		HasActiveQuest: openQuestCount(c) > 0,
	}
}

// values returns the signals in signalNames order.
func (s analyticsSignals) values() []bool {
	return []bool{s.HasDM, s.HasCharacters, s.HasNextSession, s.HasActiveQuest}
}

// score sums the weight of every satisfied signal.
func (s analyticsSignals) score() int {
	total := 0
	if s.HasDM {
		total += readinessDM
	}
	if s.HasCharacters {
		total += readinessCharacters
	}
	if s.HasNextSession {
		total += readinessNextSession
	}
	if s.HasActiveQuest {
		total += readinessActiveQuest
	}
	return total
}

// missing lists the names of the unsatisfied signals, always as a non-nil
// slice so the JSON field is [] rather than null.
func (s analyticsSignals) missing() []string {
	out := []string{}
	for i, ok := range s.values() {
		if !ok {
			out = append(out, signalNames[i])
		}
	}
	return out
}

// openQuestCount counts quests still being worked on. "Open" means active: a
// completed quest is finished and a blocked one is not something the party can
// currently pursue. Callers must hold campaigns.mu.
func openQuestCount(c *campaign) int {
	n := 0
	for _, q := range c.Quests {
		if q.Status == questActive {
			n++
		}
	}
	return n
}

// friendlyNPCCount counts NPCs with a positive disposition, matching the
// tally the relationships endpoint reports. Callers must hold campaigns.mu.
func friendlyNPCCount(c *campaign) int {
	n := 0
	for _, npc := range c.NPCs {
		if npc.friendly() {
			n++
		}
	}
	return n
}

type analyticsSummaryResponse struct {
	CampaignID        string `json:"campaign_id"`
	ReadinessScore    int    `json:"readiness_score"`
	OpenQuests        int    `json:"open_quests"`
	FriendlyNPCs      int    `json:"friendly_npcs"`
	ScheduledSessions int    `json:"scheduled_sessions"`
	InventoryItems    int    `json:"inventory_items"`
}

type riskReportRequest struct {
	IncludeZeroes *bool `json:"include_zeroes"`
}

type riskReportResponse struct {
	CampaignID string           `json:"campaign_id"`
	RiskLevel  string           `json:"risk_level"`
	Missing    []string         `json:"missing"`
	Signals    analyticsSignals `json:"signals"`
}

// ---------- GET /v1/campaigns/{id}/analytics/summary ----------

func handleAnalyticsSummary(w http.ResponseWriter, r *http.Request) {
	campaigns.mu.Lock()
	defer campaigns.mu.Unlock()
	c, ok := campaigns.campaigns[r.PathValue("id")]
	if !ok {
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	writeJSON(w, http.StatusOK, analyticsSummaryResponse{
		CampaignID:        c.ID,
		ReadinessScore:    campaignSignals(c).score(),
		OpenQuests:        openQuestCount(c),
		FriendlyNPCs:      friendlyNPCCount(c),
		ScheduledSessions: len(c.Sessions),
		InventoryItems:    len(c.Inventory),
	})
}

// ---------- POST /v1/campaigns/{id}/analytics/risk-report ----------

func handleAnalyticsRiskReport(w http.ResponseWriter, r *http.Request) {
	// The body carries only presentation options, so an absent one is fine;
	// a malformed one is still rejected.
	var req riskReportRequest
	if err := decodeBody(r, &req); err != nil && !errors.Is(err, io.EOF) {
		writeError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}

	campaigns.mu.Lock()
	defer campaigns.mu.Unlock()
	c, ok := campaigns.campaigns[r.PathValue("id")]
	if !ok {
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}

	// include_zeroes asks for the zero-valued signals to be spelled out. The
	// signals object already reports all four either way, so the flag only
	// documents intent; `missing` names the false ones unconditionally.
	signals := campaignSignals(c)
	writeJSON(w, http.StatusOK, riskReportResponse{
		CampaignID: c.ID,
		RiskLevel:  riskLevel(signals),
		Missing:    signals.missing(),
		Signals:    signals,
	})
}

// riskLevel grades a campaign by how many readiness signals it is missing.
func riskLevel(s analyticsSignals) string {
	switch len(s.missing()) {
	case 0:
		return riskLow
	case 1, 2:
		return riskMedium
	default:
		return riskHigh
	}
}
